apps/api/src/app/shared/framework/idempotency.interceptor.ts

```diff
@@ -0,0 +1,188 @@
+// idempotency.interceptor.ts
+
+import {
+  Injectable,
+  NestInterceptor,
+  ExecutionContext,
+  CallHandler,
+  Logger,
+  ConflictException,
+  HttpException,
+  InternalServerErrorException,
+} from '@nestjs/common';
+import { CacheService } from '@novu/application-generic';
+import { Observable, of, throwError } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+import { createHash } from 'crypto';
+
+const LOG_CONTEXT = 'IdempotencyInterceptor';
+const IDEMPOTENCY_CACHE_TTL = 60 * 60 * 6; //6h
+const IDEMPOTENCY_PROGRESS_TTL = 60 * 5; //5min
+
+const HEADER_KEYS = {
+  IDEMPOTENCY: 'idempotency-key',
+  RETRY_AFTER: 'retry-after',
+  IDEMPOTENCY_CONFLICT: 'x-idempotency-conflict',
+};
+
+enum ReqStatusEnum {
+  PROGRESS = 'in-progress',
+  SUCCESS = 'success',
+  ERROR = 'error',
+}
+
+@Injectable()
+export class IdempotencyInterceptor implements NestInterceptor {
+  constructor(private readonly cacheService: CacheService) {}
+
+  async intercept(context: ExecutionContext, next: CallHandler): Promise<Observable<any>> {
+    const request = context.switchToHttp().getRequest();
+    const idempotencyKey = request.headers[HEADER_KEYS.IDEMPOTENCY];
+    const isEnabled = process.env.API_IDEMPOTENCY_ENABLED == 'true';
+    if (!isEnabled || !idempotencyKey || !['post', 'patch'].includes(request.method.toLowerCase())) {
+      return next.handle();
+    }
+    const { organizationId } = request.user || {};
+    const env = process.env.NODE_ENV;
+    const cacheKey = `${env}-${organizationId}-${idempotencyKey}`;
+
+    try {
+      const hash = createHash('blake2s256');
+      hash.update(Buffer.from(JSON.stringify(request.body)));
+      const bodyHash = hash.digest('hex');
+      //if 1st time we are seeing the request, marks the request as in-progress if not, does nothing
+      const isNewReq = await this.setCache(
+        cacheKey,
+        { status: ReqStatusEnum.PROGRESS, bodyHash },
+        IDEMPOTENCY_PROGRESS_TTL,
+        true
+      );
+      // Check if the idempotency key is in the cache
+      if (isNewReq) {
+        return next.handle().pipe(
+          map(async (response) => {
+            const httpResponse = context.switchToHttp().getResponse();
+            const statusCode = httpResponse.statusCode;
+
+            // Cache the success response and return it
+            await this.setCache(
+              cacheKey,
+              { status: ReqStatusEnum.SUCCESS, bodyHash, statusCode: statusCode, data: response },
+              IDEMPOTENCY_CACHE_TTL
+            );
+            Logger.verbose(`cached the success response for idempotency`, LOG_CONTEXT);
```

Comment:
i think the idempotency key should be logged as well

Reply:
I would agree

Reply:
i have made the changes.